python/ray/dag/compiled_dag_node.py

```diff
@@ -847,7 +847,7 @@ def _preprocess(self) -> None:
                 task.dag_node, InputAttributeNode
             ):
                 continue
-            if len(task.downstream_node_idxs) == 0:
+            if len(task.downstream_node_idxs) == 0 and task.dag_node.is_output_node:
```

Comment:
The function preprocess is a bit tricky. If we don't explicitly specify `MultiOutputNode`, this function will create a new `MultiOutputNode` with the DAG node and `output_task_idx`, and then call preprocess again. As a result, in the second round of `preprocess`, the `len(task.downstream_node_idxs)` of the node with `is_output_node` will change from 0 to 1 because it now has a new downstream node, `MultiOutputNode`.

Reply:
can we just check when we first compile, if the current node is not multi output node, we append it automtaically and just run preproess once?

Reply:
Updated: https://github.com/ray-project/ray/pull/47261/commits/1a186858778b7410cd87ebfb3477ca73c1120404